File: lfric/infrastructure/integration-test/configuration_test/configuration_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
##############################################################################
# (c) Crown copyright 2017 Met Office. All rights reserved.
# The file LICENCE, distributed with this code, contains details of the terms
# under which the code may be used.
##############################################################################
'''
Exercise namelist loading capabilities.
'''
from __future__ import print_function

from __future__ import absolute_import
import collections
import re
import sys

from testframework import MpiTest, TestEngine, TestFailed
from six.moves import range


##############################################################################
class configuration_test(MpiTest):
    '''
    Tests the ability to load namelists in an MPI environment.
    '''
    # pylint: disable=invalid-name

    _INJECT = 'configuration_test.nml'

    def __init__(self):
        super(configuration_test, self).__init__([sys.argv[1], self._INJECT],
                                                 processes=4)
        self._precision = 0.0005

    def test(self, return_code, out, err):
        # pylint: disable=too-many-locals

        if return_code != 0:
            print('Standard out: {out}'.format(out=out), file=sys.stderr)
            print('Standard error: {err}'.format(err=err), file=sys.stderr)
            message = 'Unexpected failure of test executable: {code}'
            raise TestFailed(message.format(code=return_code))

        expected = {'a_dim': 2,
                    'angle_deg': 7.998,
                    'angle_rad': 0.1396,
                    'an_enum': 'second',
                    'bounded_array_local_dim': [0.3, 0.4],
                    'bounded_array1_namelist_dim': [0.3, 0.4, 0.5],
                    'bounded_array2_namelist_dim': [0.5, 0.6, 0.7],
                    'bounded_array_source_dim': [0.1, 0.6, 0.7, 0.9],
                    'closed_array': [0.2, 0.3, 0.4],
                    'open_array': [1, 2, 3, 4, 5],
                    'some_string': 'chocolate teapot',
                    'whole_number': 13}

        variable_pattern = re.compile(r'(\d+)\s+(.+?)\s*:\s*(.+)\s*')
        list_pattern = re.compile(r'(\'.*?\'|".*?"|[^ ]+)')

        seen = collections.defaultdict(list)

        for process in range(0, 4):
            with open('result.{0}.txt'.format(process), 'rt') as handle:
                contents = handle.readlines()

            for line in contents:
                match = variable_pattern.match(line)
                if match:
                    rank = int(match.group(1))
                    variable = match.group(2).strip()
                    seen[rank].append(variable)
                    value = match.group(3).strip()

                    if rank != process:
                        message = 'Found output for rank {0} in file from {1}'
                        raise TestFailed(message.format(rank, process))

                    list_match = list_pattern.findall(value)
                    if isinstance(list_match, list):
                        value = []
                    for item in list_match:
                        if item.startswith("'") or item.startswith('"'):
                            value.append(item[1:-1])
                        else:
                            value.append(item)

                    if variable not in expected:
                        message = 'Found unexpected variable "{variable}"' \
                                  ' in output'
                        raise TestFailed(message.format(variable=variable))

                    self._check(variable, expected[variable], value)

                additions = set(seen[0]) - set(expected.keys())
                if len(additions) > 0:
                    message = 'Addition variables found: {adds}'
                    raise TestFailed(message.format(adds=', '.join(additions)))

        if list(seen.keys()) != [0, 1, 2, 3]:
            message = 'Incorrect ranks: ' + str(list(seen.keys()))
            raise TestFailed(message)

        return 'One of each configuration type loaded'

    def _check(self, name, expected, found):
        # pylint: disable=consider-using-enumerate
        if not isinstance(expected, list):
            expected = [expected]
        if not isinstance(found, list):
            found = [found]

        if len(found) != len(expected):
            message = 'Expected variable "{variable}" to hold "{expect}"' \
                      ' but found "{found}", different number of values.'
            raise TestFailed(message.format(variable=name,
                                            expect=expected,
                                            found=found))

        for index in range(len(expected)):
            if isinstance(expected[index], float):
                maybe_zero = abs(float(found[index]) - expected[index])
                if maybe_zero > self._precision:
                    message = 'Expected index {index} of variable' \
                              ' "{variable}" to be within {precision} of' \
                              ' {expected} but found {found}'
                    raise TestFailed(message.format(variable=name,
                                                    index=index,
                                                    precision=self._precision,
                                                    expected=expected[index],
                                                    found=found[index]))
            else:  # expected[index] is not float
                if found[index] != str(expected[index]):
                    message = 'Expected index {index} of variable' \
                              '"{variable}" to hold "{expect}" but found' \
                              ' "{found}"'
                    raise TestFailed(message.format(variable=name,
                                                    index=index,
                                                    expect=expected[index],
                                                    found=found[index]))


##############################################################################
if __name__ == '__main__':
    TestEngine.run(configuration_test())

File: lfric/infrastructure/integration-test/configuration_test/configuration_test.f90
!-----------------------------------------------------------------------------
! (C) Crown copyright 2017 Met Office. All rights reserved.
! The file LICENCE, distributed with this code, contains details of the terms
! under which the code may be used.
!-----------------------------------------------------------------------------

program configuration_test

  use, intrinsic :: iso_fortran_env, only : error_unit
  use mpi_mod,                     only : initialise_comm, store_comm,       &
                                          finalise_comm,                     &
                                          get_comm_rank
  use one_of_each_test_config_mod, only : key_from_an_enum,                  &
                                      postprocess_one_of_each_test_namelist, &
                                          read_one_of_each_test_namelist,    &
                                          a_dim,                             &
                                          angle_deg,                         &
                                          angle_rad,                         &
                                          an_enum,                           &
                                          bounded_array_local_dim,           &
                                          bounded_array1_namelist_dim,       &
                                          bounded_array2_namelist_dim,       &
                                          bounded_array_source_dim,          &
                                          closed_array,                      &
                                          max_array_size,                    &
                                          open_array,                        &
                                          some_string,                       &
                                          whole_number

   use another_list_config_mod, only : postprocess_another_list_namelist,    &
                                       read_another_list_namelist,           &
                                       some_other_dim

  implicit none

  integer,      parameter :: file_unit = 13
  character(*), parameter :: filename = 'configuration_test.nml'

  integer       :: comm
  integer       :: rank
  integer       :: condition
  character(20) :: result_filename
  character(40) :: format_string

  ! Initialse mpi and create the default communicator: mpi_comm_world
  call initialise_comm(comm)

  ! Save lfric's part of the split communicator for later use
  call store_comm(comm)

  rank = get_comm_rank()

  open( file_unit, file=filename, iostat=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to open file: ",A)' ) filename
    stop 3
  end if

  call read_another_list_namelist( file_unit ,rank )
  REWIND(file_unit)
  call read_one_of_each_test_namelist( file_unit ,rank )

  close( file_unit, iostat=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to close file: ", A)' ) filename
    stop 4
  end if

  call postprocess_another_list_namelist
  call postprocess_one_of_each_test_namelist

  write( result_filename, '("result.", I0, ".txt")' ) rank
  open( file_unit, file=result_filename, iostat=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to open file: ",A)' ) result_filename
    stop 5
  end if

  write( file_unit, '(I0, A, I0)' ) rank, ' a_dim: ', a_dim
  write( file_unit, '(I0, " angle_deg: ", E14.7)' ) rank, angle_deg
  write( file_unit, '(I0, " angle_rad: ", E14.7)' ) rank, angle_rad
  write( file_unit, '(I0, " an_enum: ", A)' ) rank, &
                                              trim(key_from_an_enum( an_enum ))

  write( format_string, '(A,I0,A)') '(I0, A, ', max_array_size,  'E14.7)'
  write( file_unit, format_string ) &
    rank, ' bounded_array_local_dim:', bounded_array_local_dim
  write( file_unit, format_string ) &
    rank, ' bounded_array1_namelist_dim:', bounded_array1_namelist_dim
  write( file_unit, format_string ) &
    rank, ' bounded_array2_namelist_dim:', bounded_array2_namelist_dim
  write( file_unit, format_string ) &
    rank, ' bounded_array_source_dim:', bounded_array_source_dim
  write( file_unit, format_string ) rank, ' closed_array: ', closed_array

  write( format_string, &
         '("(I0, "" open_array: "", ", I0, "I3)")' ) max_array_size
  write( file_unit, format_string ) rank, open_array
  write( file_unit, '(I0, " some_string: ''", A, "''")' ) rank, &
                                                           trim(some_string)
  write( file_unit, '(I0, " whole_number: ", I0)' ) rank, whole_number

  close( file_unit, iostat=condition )
  if (condition /= 0) then
    write( error_unit, '("Failed to close file: ", A)' ) result_filename
    stop 6
  end if

  call finalise_comm()

end program configuration_test
